namerd/storage/zk/src/main/scala/com/twitter/finagle/serverset2/buoyant/ZkDtabStore.scala

```diff
@@ -0,0 +1,127 @@
+package com.twitter.finagle.serverset2.buoyant
+
+import com.twitter.finagle.Dtab
+import com.twitter.finagle.serverset2.client.KeeperException.{BadVersion, NoNode}
+import com.twitter.finagle.serverset2.client._
+import com.twitter.finagle.serverset2.{RetryStream, Zk2Resolver, ZkSession => FZkSession}
+import com.twitter.finagle.stats.DefaultStatsReceiver
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.io.Buf
+import com.twitter.logging.Logger
+import com.twitter.util._
+import io.buoyant.namerd.DtabStore.{DtabNamespaceAlreadyExistsException, DtabNamespaceDoesNotExist, DtabVersionMismatchException}
+import io.buoyant.namerd.{DtabStore, VersionedDtab}
+import java.nio.ByteBuffer
+
+/**
+ * A DtabStore which stores dtabs in ZooKeeper.  Dtabs are stored, Utf8
+ * encoded, in the ZooKeeper node /dtabs/<ns>.
+ */
+class ZkDtabStore(
+  hosts: String,
+  zkPrefix: String,
+  sessionTimeout: Option[Duration]
+) extends DtabStore {
+
+  private[this] val log = Logger()
+  private[this] val retryStream = RetryStream()
+  private[this] val stats = DefaultStatsReceiver.scope("zkclient").scope(Zk2Resolver.statsOf(hosts))
+
+  private[this] implicit val timer = DefaultTimer.twitter
+  private[this] val builder = ClientBuilder()
+    .hosts(hosts)
+    .sessionTimeout(sessionTimeout.getOrElse(FZkSession.DefaultSessionTimeout))
+    .statsReceiver(stats)
+
+  private[this] val zkSession = new ZkSession(
+    retryStream,
+    retryStream,
+    () => builder.writer(),
+    stats
+  )
+  private[this] def actOf[T](go: ZooKeeperRW => Future[Watched[T]]): Activity[T] =
+    zkSession.watchedOperation(go(zkSession.zk))
+
+  private[this] val actNs = actOf(_.getChildrenWatch(zkPrefix)).map(_.children.toSet)
+
+  def list(): Future[Set[String]] = actNs.values.toFuture.flatMap(Future.const)
+
+  def create(ns: String, dtab: Dtab): Future[Unit] = {
+    val path = s"$zkPrefix/$ns"
+    log.info(s"Attempting to create dtab at $path")
+
+    zkSession.zk.create(
+      path,
+      Some(Buf.Utf8(dtab.show)),
+      Seq(Data.ACL.AnyoneAllUnsafe),
+      CreateMode.Persistent
+    ).rescue {
+        case KeeperException.NodeExists(_) =>
+          Future.exception(new DtabNamespaceAlreadyExistsException(ns))
+      }.unit
```

Comment:
Doesn't really matter either way in this case.  This ordering makes sense to me because `zk.create` is the fallible operation and I like to have the rescue directly on the operation that it rescues from.  The `.unit` call is just to discard the return value since it's not needed to satisfy the method signature so it makes sense to me to have it last.